include/evaluators/blas1_tree_evaluator.hpp

```diff
@@ -41,100 +41,42 @@ namespace blas {
 
-  size_t blqS = 256;
-  size_t grdS = 512;
-
   Evaluator(Expression &expr)
-      : l(Evaluator<LHS, Device>(expr.l)), r(Evaluator<RHS, Device>(expr.r)) {
-    setBlocksize(256);
-    setGridsize(512);
-  }
+      : l(Evaluator<LHS, Device>(expr.l)), r(Evaluator<RHS, Device>(expr.r)) {}
 
   size_t getSize() const { return r.getSize(); }
   cont_type *data() { return l.data(); }
 
-  void setBlocksize(size_t blocksize) { blqS = blocksize; }
-
-  void setGridsize(size_t gridsize) { grdS = gridsize; }
-
   void reduce(Device &dev);
 
   bool eval_subexpr_if_needed(cont_type *cont, Device &dev) {
     l.eval_subexpr_if_needed(NULL, dev);
     r.eval_subexpr_if_needed(NULL, dev);
-    reduce(dev);
-    return true;
-  }
-
-  value_type eval(size_t i) {
-    size_t vecS = r.getSize();
-
-    size_t frs_thrd = 2 * blqS * i;
-    size_t lst_thrd = ((frs_thrd + blqS) > vecS) ? vecS : (frs_thrd + blqS);
-    // Reduction across the grid
-    value_type val = dev_functor::init(r);
-    for (size_t j = frs_thrd; j < lst_thrd; j++) {
-      value_type local_val = dev_functor::init(r);
-      for (size_t k = j; k < vecS; k += 2 * grdS) {
-        local_val = dev_functor::eval(local_val, r.eval(k));
-        if (k + blqS < vecS) {
-          local_val = dev_functor::eval(local_val, r.eval(k + blqS));
-        }
-      }
-      // Reduction inside the block
-      val = dev_functor::eval(val, local_val);
+    if (cont)){
+      m_result = cont;
     }
-    return l.eval(i) = val;
-  }
-
-  template <typename sharedT>
-  value_type eval(sharedT scratch, cl::sycl::nd_item<1> ndItem) {
-    size_t localid = ndItem.get_local(0);
-    size_t localSz = ndItem.get_local_range(0);
-    size_t groupid = ndItem.get_group(0);
-
-    size_t vecS = r.getSize();
-    size_t frs_thrd = 2 * groupid * localSz + localid;
-
-    // Reduction across the grid
-    value_type val = dev_functor::init(r);
-    for (size_t k = frs_thrd; k < vecS; k += 2 * grdS) {
-      val = dev_functor::eval(val, r.eval(k));
-      if ((k + blqS < vecS)) {
-        val = dev_functor::eval(val, r.eval(k + blqS));
-      }
+    else {
+      m_result = dev.allocate(out_size);
     }
-
-    scratch[localid] = val;
-    // This barrier is mandatory to be sure the data is on the shared memory
-    ndItem.barrier(cl::sycl::access::fence_space::local_space);
-
-    // Reduction inside the block
-    for (size_t offset = localSz >> 1; offset > 0; offset >>= 1) {
-      if (localid < offset) {
-        scratch[localid] =
-            dev_functor::eval(scratch[localid], scratch[localid + offset]);
-      }
-      // This barrier is mandatory to be sure the data are on the shared memory
-      ndItem.barrier(cl::sycl::access::fence_space::local_space);
-    }
-    if (localid == 0) {
-      l.eval(groupid) = scratch[localid];
-    }
-    return l.eval(groupid);
+    FullReducer<Self, Device>::run(this, dev, *result);
+    return true;
   }
 
+  value_type eval(size_t i) { return m_result[i]; }
```

Comment:
The assignment is not fixed yet, so it makes sense to figure out whether it is value or reference when it starts to matter.